### PR TITLE
Functionalities to incorporate rotational averaging in i-PI

### PR DESCRIPTION
**Pull request by ceriottm**
The idea is to simplify the use (or at least the benchmarking) of unconstrained models by performing i-PI side the rotational averaging.

### CODE REVIEW

File: ipi/engine/forcefields.py [hunk 1]
```diff
@@ -1371,6 +1374,162 @@ def poll(self):
...
+            rot_atoms = atoms.clone()
+            # NB we need generic cell orientation
+            rot_cell = GenericCell(R @ dstrip(cell.h).copy())
+            rot_atoms.q[:] = (dstrip(rot_atoms.q).reshape(-1, 3) @ R.T).flatten()
+            rots.append((R, w))
+
+            ffh.append(super(FFRotations, self).queue(rot_atoms, rot_cell, reqid))
+
+            if self.inversion:
+                # also add a "flipped cell" to the evaluation list
```
Comment (sirmarcel) [inline ipi/engine/forcefields.py:1435]: "flipped rotation", no?

File: ipi/engine/forcefields.py [hunk 2]
```diff
@@ -1371,6 +1374,162 @@ def poll(self):
...
+        if self.random:
+            R_random = random_rotation(self.prng, improper=True)
+        else:
+            R_random = np.eye(3)
+
+        for R, w in self._rotations:
+            R = R @ R_random
+            rot_atoms = atoms.clone()
+            # NB we need generic cell orientation
+            rot_cell = GenericCell(R @ dstrip(cell.h).copy())
```
Comment (sirmarcel) [inline ipi/engine/forcefields.py:1428]: What does `cell.h` do?

File: ipi/engine/forcefields.py [hunk 3]
```diff
@@ -1371,6 +1374,162 @@ def poll(self):
...
+            rot_atoms.q[:] = (dstrip(rot_atoms.q).reshape(-1, 3) @ R.T).flatten()
```
Comment (sirmarcel) [inline ipi/engine/forcefields.py:1429]: This seems like it does the correct thing of doing `R @ pos` where `pos` is column-wise the positions. Just checking.

**Review by sirmarcel — COMMENTED** [review]
Looks fine overall, added a first round of comments for some details.

File: ipi/utils/mathtools.py [hunk 4]
```diff
@@ -521,3 +521,111 @@ def f(w):
...
+    # Calculate weights using the formula
+    weights = 2 / ((1 - roots**2) * (legendre_poly_deriv(roots) ** 2))
+
+    return roots, weights
+
+
+def get_rotation_quadrature(L):
+    if L == 1:
+        # returns the identity (for some reason this algo below generates a different rotation)
+        return [(np.eye(3), 2.0)]
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply ipi/utils/mathtools.py:603]: Seems it's how it's implemented in the reference paper. I'm not convinced this is the most efficient O3 grid, but for the moment I'd go with it and stick to the implementation.

File: ipi/engine/forcefields.py (hunk 3 above)
Comment (ceriottm) [reply ipi/engine/forcefields.py:1429]: yes, it does the right thing with q being stored with x,y,z in rows

File: ipi/engine/forcefields.py (hunk 2 above)
Comment (ceriottm) [reply ipi/engine/forcefields.py:1428]: `cell` is an object, `cell.h` is the actual cell matrix.

File: ipi/engine/forcefields.py [hunk 5]
```diff
@@ -1371,6 +1374,162 @@ def poll(self):
...
+
+            ffh.append(super(FFRotations, self).queue(rot_atoms, rot_cell, reqid))
+
+            if self.inversion:
+                # also add a "flipped cell" to the evaluation list
+                R = R * -1
+                rot_cell = GenericCell(R @ dstrip(cell.h).copy())
+                rot_atoms = atoms.clone()
+                rot_atoms.q[:] = (dstrip(rot_atoms.q).reshape(-1, 3) @ R.T).flatten()
+                ffh.append(super(FFRotations, self).queue(rot_atoms, rot_cell, reqid))
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply ipi/engine/forcefields.py:1440]: yes, that's correct. this was a bug, so :+1:

File: ipi/engine/forcefields.py [hunk 6]
```diff
@@ -1371,6 +1374,162 @@ def poll(self):
...
+        rot_handles = r["ff_handles"].copy()
+        rots = r["rots"].copy()
+
+        # Gathers the forcefield energetics and extras
+        pots = []
+        frcs = []
+        virs = []
+        xtrs = []
+        quad_w = []
+        for ff_r, (R, w) in zip(rot_handles, rots):
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply ipi/engine/forcefields.py:1480]: yes, rotation matrices were missing. now we add them, so all should be good.

File: ipi/engine/forcefields.py [hunk 7]
```diff
@@ -1371,6 +1374,162 @@ def poll(self):
...
+            quad_w.append(w)
+
+        quad_w = np.array(quad_w)
+        pots = np.array(pots)
+        frcs = np.array(frcs).reshape(len(frcs), -1)
+        virs = np.array(virs).reshape(-1, 3, 3)
+
+        # Computes the mean energetics (using the quadrature weights)
+        mean_pot = np.mean(pots * quad_w, axis=0) / quad_w.mean()
+        mean_frc = np.mean(frcs * quad_w[:, np.newaxis], axis=0) / quad_w.mean()
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply ipi/engine/forcefields.py:1495]: no, actually.

File: ipi/engine/forcefields.py [hunk 8]
```diff
@@ -1371,6 +1374,162 @@ def poll(self):
...
+        self.grid = grid
+
+        self._rotations = get_rotation_quadrature(self.grid)
+
+    def queue(self, atoms, cell, reqid=-1):
+        # launches requests for all of the rotations FF objects
+        ffh = []
+        rots = []
+        if self.random:
+            R_random = random_rotation(self.prng, improper=True)
```
[the remark this answers is not in the document]
Comment (ceriottm) [reply ipi/engine/forcefields.py:1420]: Seems one could inversion + random for instance

**Comment by ceriottm** [comment]
It'd be good if you could somewhat test it - maybe in conjunction with work on the o3-averaging recipe

**Comment by ceriottm** [comment]
This has been tested extensively by both me and @DavideTisi. We're happy and ready to merge.

**Review by DavideTisi — APPROVED** [review]
As Michele said we tested and it works as intended

**Pull request merged**